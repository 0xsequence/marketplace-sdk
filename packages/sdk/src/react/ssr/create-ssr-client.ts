import type { QueryClient } from '@tanstack/react-query';
import { type State, cookieToInitialState } from 'wagmi';
import type { SdkConfig } from '../../types/sdk-config';
import { createWagmiConfig } from '../_internal/wagmi/create-config';
import { marketplaceConfigOptions } from '../hooks/options/marketplaceConfigOptions';

type InitSSRClientArgs = {
	cookie: string;
	config: SdkConfig;
	queryClient: QueryClient;
};

type InitialState = { wagmi?: State };

const marketplaceConfig = async (
	config: SdkConfig,
	queryClient: QueryClient,
) => {
	const configOptions = marketplaceConfigOptions(config);
	return queryClient.fetchQuery(configOptions);
};

const initialState = async (args: InitSSRClientArgs): Promise<InitialState> => {
	const marketConfig = await marketplaceConfig(args.config, args.queryClient);
	const wagmiConfig = createWagmiConfig(marketConfig, args.config, true);
	return { wagmi: cookieToInitialState(wagmiConfig, args.cookie) };
};

export const createSSRClient = (args: InitSSRClientArgs) => {
	const getMarketplaceConfig = async () =>
		marketplaceConfig(args.config, args.queryClient);
	const getInitialState = async () => initialState(args);
	const config = args.config;
	return { getInitialState, getMarketplaceConfig, config };
};
