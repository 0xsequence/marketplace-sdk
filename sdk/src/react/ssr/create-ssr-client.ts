import { cookieToInitialState, type State } from 'wagmi';
import type { SdkConfig } from '../../types/sdk-config';
import { getQueryClient } from '../_internal/api/get-query-client';
import { createWagmiConfig } from '../_internal/wagmi/create-config';
import { marketplaceConfigOptions } from '../queries/market/marketplaceConfig';

type InitSSRClientArgs = {
	cookie: string;
	config: SdkConfig;
};

type InitialState = { wagmi?: State };

const marketplaceConfig = async (config: SdkConfig) => {
	const configOptions = marketplaceConfigOptions(config);
	return getQueryClient().fetchQuery(configOptions);
};

const initialState = async (args: InitSSRClientArgs): Promise<InitialState> => {
	const marketConfig = await marketplaceConfig(args.config);
	const wagmiConfig = createWagmiConfig(marketConfig, args.config, true);
	return { wagmi: cookieToInitialState(wagmiConfig, args.cookie) };
};

export const createSSRClient = (args: InitSSRClientArgs) => {
	const getMarketplaceConfig = async () => marketplaceConfig(args.config);
	const getInitialState = async () => initialState(args);
	const config = args.config;
	return { getInitialState, getMarketplaceConfig, config };
};
