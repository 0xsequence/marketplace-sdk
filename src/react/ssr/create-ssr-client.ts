import { type State, cookieToInitialState } from 'wagmi';
import type { SdkConfig } from '../../types/sdk-config';
import { getQueryClient } from '../_internal/api/get-query-client';
import { createWagmiConfig } from '../_internal/wagmi/create-config';
import { marketplaceConfigOptions } from '../hooks/options/marketplaceConfigOptions.ts';

type InitSSRClientArgs = {
	cookie: string;
	config: SdkConfig;
};

type InitialState = { wagmi?: State };

const marketplaceConfig = async (config: SdkConfig) => {
	const queryClient = getQueryClient();
	const configOptions = marketplaceConfigOptions(config);
	return queryClient.fetchQuery(configOptions);
};

const initialState = async (args: InitSSRClientArgs): Promise<InitialState> => {
	const marketConfig = await marketplaceConfig(args.config);
	const wagmiConfig = createWagmiConfig(marketConfig, args.config, true);
	return { wagmi: cookieToInitialState(wagmiConfig, args.cookie) };
};

export const createSSRClient = (args: InitSSRClientArgs) => {
	const getMarketplaceConfig = async () => marketplaceConfig(args.config);
	const getInitialState = async () => initialState(args);
	const config = args.config;
	return { getInitialState, getMarketplaceConfig, config };
};
