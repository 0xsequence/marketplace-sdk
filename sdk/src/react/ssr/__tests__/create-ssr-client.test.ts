import { describe, expect, it } from 'vitest';
import type { SdkConfig } from '../../../types';
import { createSSRClient } from '../create-ssr-client';

const wagmiCookie =
	'wagmi.store={"state":{"connections":{"__type":"Map","value":[]},"chainId":1,"current":null},"version":2}; ';

const config = {
	projectId: 'test-project',
	projectAccessKey: 'test-key',
} satisfies SdkConfig;

const client = createSSRClient({
	cookie: wagmiCookie,
	config,
});

describe('createSSRClient', () => {
	it('should fetch marketplace config successfully', async () => {
		const marketplaceConfig = await client.getMarketplaceConfig();

		expect(marketplaceConfig).toMatchSnapshot();
	});

	it('should get initial state configuration', async () => {
		const state = await client.getInitialState();

		expect(state).toMatchInlineSnapshot(`
			{
			  "wagmi": {
			    "chainId": 1,
			    "connections": Map {},
			    "current": null,
			  },
			}
		`);
	});

	it('should preserve provided config in the client', () => {
		expect(client.config).toEqual(config);
	});
});
