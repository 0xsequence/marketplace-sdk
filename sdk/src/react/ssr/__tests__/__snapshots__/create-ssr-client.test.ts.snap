// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`createSSRClient > should fetch marketplace config successfully 1`] = `
{
  "cssString": "
  .marketplace-theme {
    --primary-color: #000000;
  }
",
  "manifestUrl": "https://api.sequence.build/marketplace/test-project/manifest.json",
  "market": {
    "bannerUrl": "",
    "collections": [
      {
        "bannerUrl": "",
        "chainId": 1,
        "currencyOptions": [
          "0x0000000000000000000000000000000000000000",
          "0xA0b86991c6218b36c1d19D4a2e9Eb0cE3606eB48",
        ],
        "destinationMarketplace": "sequence_marketplace_v2",
        "feePercentage": 3.5,
        "filterSettings": {
          "exclusions": [
            {
              "condition": "SPECIFIC_VALUE",
              "key": "Type",
              "value": "Sample",
            },
          ],
          "filterOrder": [
            "Type",
            "Rarity",
          ],
        },
        "isLAOSERC721": undefined,
        "itemsAddress": "0x0000000000000000000000000000000000000000",
        "marketplaceType": "MARKET",
      },
      {
        "bannerUrl": "https://example.com/collection-banner.png",
        "chainId": 137,
        "currencyOptions": [
          "0x0000000000000000000000000000000000000000",
        ],
        "destinationMarketplace": "opensea",
        "feePercentage": 2.5,
        "filterSettings": {
          "exclusions": [
            {
              "condition": "ENTIRE_KEY",
              "key": "Category",
            },
            {
              "condition": "SPECIFIC_VALUE",
              "key": "Level",
              "value": "Legendary",
            },
          ],
          "filterOrder": [
            "Category",
            "Level",
            "Element",
          ],
        },
        "isLAOSERC721": undefined,
        "itemsAddress": "0x1234567890123456789012345678901234567890",
        "marketplaceType": "MARKET",
      },
    ],
    "enabled": true,
    "ogImage": undefined,
    "title": "Test Marketplace",
  },
  "projectId": NaN,
  "settings": {
    "accessKey": undefined,
    "faviconUrl": "https://example.com/favicon.png",
    "fontUrl": undefined,
    "logoUrl": "https://example.com/logo.png",
    "publisherId": "test-publisher",
    "socials": {
      "discord": "https://discord.com/test",
      "instagram": "https://instagram.com/test",
      "tiktok": "",
      "twitter": "https://twitter.com/test",
      "website": "",
      "youtube": "",
    },
    "title": "Test Marketplace",
    "walletOptions": {
      "connectors": [
        "coinbase",
        "walletconnect",
      ],
      "includeEIP6963Wallets": true,
      "oidcIssuers": {},
      "walletType": "UNIVERSAL",
    },
  },
  "shop": {
    "bannerUrl": "",
    "collections": [
      {
        "bannerUrl": "",
        "chainId": 80002,
        "filterSettings": undefined,
        "isLAOSERC721": false,
        "itemsAddress": "0xf2ea13ce762226468deac9d69c8e77d291821676",
        "marketplaceType": "SHOP",
        "saleAddress": "0x30131575129ee043f9c7409ca599bfd8ffe1b4e0",
      },
    ],
    "enabled": true,
    "ogImage": undefined,
    "title": "",
  },
}
`;
